learning_resources_search/api.py

```diff
@@ -517,9 +531,101 @@ def execute_learn_search(search_params):
         )
         search = search.extra(aggs=aggregation_clauses)
 
+    return search
+
+
+def execute_learn_search(search_params):
+    """
+    Execute a learning resources search based on the query
+
+
+    Args:
+        search_params (dict): The opensearch query params returned from
+        LearningResourcesSearchRequestSerializer
+
+    Returns:
+        dict: The opensearch response dict
+    """
+
+    search = construct_search(search_params)
+
     return search.execute().to_dict()
 
 
+def subscribe_user_to_search_query(user, search_params):
+    from learning_resources_search.models import PercolateQuery
+
+    """
+    Subscribe a user to a search query
+
+
+    Args:
+        user: The User to subscribe
+        search_params (dict): The opensearch query params returned from
+        LearningResourcesSearchRequestSerializer
+
+    Returns:
+        dict: The opensearch response dict
+    """
+
+    percolate_query, _ = PercolateQuery.objects.get_or_create(
+        source_type=PercolateQuery.SEARCH_SUBSCRIPTION_TYPE,
+        original_query=adjust_original_query_for_percolate(search_params),
+    )
+    if not percolate_query.users.filter(id=user.id).exists():
+        percolate_query.users.add(user)
+    return percolate_query.query
+
+
+def unsubscribe_user_to_search_query(user, search_params):
+    from learning_resources_search.models import PercolateQuery
+
+    """
+    Unsubscribe a user to a search query
+
+
+    Args:
+        user: The User to unsubscribe
+        search_params (dict): The opensearch query params returned from
+        LearningResourcesSearchRequestSerializer
+
+    Returns:
+        dict: The opensearch response dict
+    """
+    percolate_query, _ = PercolateQuery.objects.get_or_create(
+        source_type=PercolateQuery.SEARCH_SUBSCRIPTION_TYPE,
+        original_query=adjust_original_query_for_percolate(search_params),
+    )
+    if percolate_query.users.filter(id=user.id).exists():
+        percolate_query.users.remove(user)
+    return percolate_query.query
+
+
+def user_subscribed_to_query(user, search_params):
+    from learning_resources_search.models import PercolateQuery
+
+    """
+    Check if a user is subscribed to a search query
+
+
+    Args:
+        user: The User to check
+        search_params (dict): The opensearch query params returned from
+        LearningResourcesSearchRequestSerializer
+
+    Returns:
+        bool: Whether or not the user has subscribed to the query
+    """
+    try:
+        percolate_query = PercolateQuery.objects.get(
+            source_type=PercolateQuery.SEARCH_SUBSCRIPTION_TYPE,
+            original_query=adjust_original_query_for_percolate(search_params),
```

Comment:
This seems sensitive to the order of the search parameters.

For instance, after subscribing to a query,
GET http://localhost:8063/api/v1/learning_resources_search_subscribe/?aggregations=resource_type&aggregations=topic&aggregations=offered_by&limit=10&offset=0&q=testing&resource_type=course&resource_type=program&resource_type=podcast

returns `{"is_subscribed": true}`

but if I reorder the parameters slightly:
GET http://localhost:8063/api/v1/learning_resources_search_subscribe/?aggregations=resource_type&aggregations=offered_by&aggregations=topic&limit=10&offset=0&q=testing&resource_type=course&resource_type=program&resource_type=podcast

I get `{"is_subscribed": false}`

Reply:
`PercolateQuery` stores the query as JSON, right? I have not tested this, but my guess would be that queries
```
// equivalent queries, both stored internally as { a: 1, b: "B" }
?a=1&b=B
?b=B&a=1

// NOT equivalent because `{ a: [1, 2], b: "B" } and { a: [2, 1], b: "B" }`
// are distinct JSON objects.
?a=1&a=2&b=B
?a=2&a=1&b=B
```

Maybe we could sort keys? Either via `save()` int he model, or in one of these api handlers?

Reply:
Good, catch. What I have now done is added a utility that recursively orders the params object prior to creating the query and added a test to make sure the ordering doesnt matter. It made sense to do this outside of save to avoid unintended consequences for example if someone in the future tried to lookup an item for existence only to get a mysterious duplicate error when saving.